agenthub/codeact_agent/prompt.py

```diff
@@ -0,0 +1,222 @@
+find_file <file_name> [<dir>] - finds all files with the given name in dir. If dir is not provided, searches in the current directory
+edit <start_line>:<end_line>
+<replacement_text>
+end_of_edit - replaces lines <start_line> through <end_line> (inclusive) with the given text in the open file. The replacement text is terminated by a line with only end_of_edit on it. All of the <replacement text> will be entered, so make sure your indentation is formatted properly. Python files will be checked for syntax errors after the edit. If the system detects a syntax error, the edit will not be executed. Simply try to edit the file again, but make sure to read the error message and modify the edit command you issue accordingly. Issuing the same command a second time will just lead to the same error message again.
+"""
+
+_COMMAND_DOCS = (
+    '\nApart from the standard bash commands, you can also use the following special commands in bash:\n'
+    f'{_SWEAGENT_BASH_DOCS}'
+    "Please note that THE EDIT COMMAND REQUIRES PROPER INDENTATION. If you'd like to add the line '        print(x)' you must fully write that out, with all those spaces before the code! Indentation is important and code that is not indented correctly will fail and require fixing before it can be run."
+)
+
+SYSTEM_MESSAGE = f"""A chat between a curious user and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the user's questions.
+The assistant can interact with an interactive Python (Jupyter Notebook) environment and receive the corresponding output when needed. The code should be enclosed using "<execute_ipython>" tag, for example: <execute_ipython> print("Hello World!") </execute_ipython>.
+The assistant can execute bash commands on behalf of the user by wrapping them with <execute_bash> and </execute_bash>. For example, you can list the files in the current directory by <execute_bash> ls </execute_bash>.
+The assistant should attempt fewer things at a time instead of putting too much commands OR code in one "execute" block.
+The assistant can install Python packages through bash by <execute_bash> pip install [package needed] </execute_bash> and should always import packages and define variables before starting to use them.
+The assistant should stop <execute> and provide an answer when they have already obtained the answer from the execution result.
+
+{_COMMAND_DOCS}
+
+Whenever possible, execute the code for the user using <execute_ipython> or <execute_bash> instead of providing it.
+The assistant's response should be concise, but do express their thoughts.
+Try to include one of <execute_ipython> or <execute_bash> in each of your responses, unless it is a direct answer to a question OR a message to the user.
+"""
+
+EXAMPLES = """
+--- START OF EXAMPLE ---
+
+USER: Can you create a list of numbers from 1 to 10, and create a web page to display them at port 5000?
```

Comment:
It's a bit counter-intuitive for the LLM to choose executing python code to write stuff to a file, while it has other choices like standard bash commands.

Questions:
1. Do you by intention guide the LLM to favor `<execute_ipython>` over `<execute_bash>`, or `open` + `edit` commands?
2. (just curious) If you don't provide this example, and ask the same question, would the assistant use `<execute_ipython>`?

Reply:
I think the SWE-Agent primitive is a bit odd for file creation; they would create a new file, open it, and use "edit" to dump in content (3 actions/turns in a role), which I felt is pretty weird in a sense so i randomly adopted python for writing lol.

But other than that, I think as long as they can write to file, it probably doesn't hurt too much - I'm pretty open to (and will) explore if there's a better solution without involving parsing more actions.

I have a pretty strong opinion in making agent's agent space minimalistic in a way: I hope agents can do all of their work in `<execute_bash>` and `<execute_ipython>` without producing action calls in a specific format (e.g., JSON) for parsing (which is a total pain to get it right & can constraint the agent's general capability - that's why i didn't use the `FileWriteAction` for this).

Reply:
Fair enough!